### PR TITLE
Adding support for metadata in State API Requests

### LINKED ISSUE #1909: STATE - Get State Operation should support sending metadata
The client state operations for "get" does not take a metadata dictionary. 
This makes using "partitionKey" in the cosmos state component "impossible".

[cosmos state component usage](https://github.com/dapr/components-contrib/blob/3efcb4043035742eeddbc0b182db1569ca2891d4/state/azure/cosmosdb/cosmosdb.go#L159).

### LINKED ISSUE #1967: Add optional metadata to either the GET, either bulk GET operations (or both) of the STATE API
## In what area(s)?

 /area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

## Describe the feature
Currently, both the simple POST and TRANSACTION endpoints let you specify a metadata payload in which you can define the partition key. However, when you don't specify any, Dapr will use the combination of dapr appid||key value to define the partition key value. That is all fine but if you posted a new document with a partition key value of yours, you can't retrieve it back because nor the GET, nor the BULK GET endpoints allow you to specify that partition key value.

Thanks

### PR DESCRIPTION
**Pull request by shalabhms**
# Description

Adding support for metadata in Get/Save/Delete State Requests.

## Issue reference

Closes #1909 
Closes #1967

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: dapr/proto/runtime/v1/dapr.proto [hunk 1]
```diff
@@ -132,6 +141,9 @@ message SaveStateRequest {
 
   // The array of the state key values.
   repeated common.v1.StateItem states = 2;
+
+  // The metadata which will be sent to state store components.
+  map<string,string> metadata = 3;
```
Comment (amanbha) [inline dapr/proto/runtime/v1/dapr.proto:146]: metadata at this level is not needed as StateItem has metadata, adding it here would be confusing.
When you implement [this issue about get, save parity](https://github.com/dapr/dapr/issues/1891), SaveState will become SaveBulkState at that time we can discuss if metadata needs to be moved at api level

Comment (shalabhms) [reply dapr/proto/runtime/v1/dapr.proto]: sounds good. Done .

**Comment by yaron2** [comment]
Where's the HTTP support?

**Comment by yaron2** [comment]
> Where's the HTTP support?

More specifically, in HTTP Get/BulkGet.

**Comment by shalabhms** [comment]
> Where's the HTTP support?

HTTP Get metadata support was already there. However missing was Delete one that I added now.

**Comment by yaron2** [comment]
> > Where's the HTTP support?
> 
> HTTP Get/BulkGet metadata support was already there. However missing was Delete one that I added now.

Ok, it looks like we need to update our docs about how to send metadata for `GET` requests, as this is different, not being on the body.

**Comment by shalabhms** [comment]
> > > Where's the HTTP support?
> > 
> > 
> > HTTP Get/BulkGet metadata support was already there. However missing was Delete one that I added now.
> 
> Ok, it looks like we need to update our docs about how to send metadata for `GET` requests, as this is different, not being on the body.

Ok. I have created issue https://github.com/dapr/docs/issues/767 (in 0.11 milestone ) to track this and have self assigned.

**Comment by skyao** [comment]
@shalabhms We also update the sdk to support meta in state delete. I find that in go-sdk, no metadata in delete api.  And I need this now to delete some state which metadata is required. I have summit a issue and a PR:

https://github.com/dapr/go-sdk/issues/81

**Pull request merged**